Extend list of hacks for reading non-standard complied 1d WCS fits files

hamogu opened the pull request:
In the current code, there is a hack to allow "LINEAR" as a spectral
axis. That's clearly not compliant with the FITS WCS standard as defined in
https://www.aanda.org/articles/aa/full/2006/05/aa3818-05/aa3818-05.html
and might just be motivated by the GMOS spectrum currently used as
example for this loader (@kelle ?).
This code change makes this list of keywords that will be treated as
spectral extensible and adds "Wavelength" (which is much closer to
"WAVE" as required in the standard than "LINEAR").

Tests and docs are still missing, but those are easy to add once we agree on this approach.

nmearl commented on specutils/wcs/adapters/fitswcs_adapter.py:
```diff
@@ -71,17 +81,18 @@ def spec_axis(self):
-            except ValueError:
+        if (self._spec_axis < 0) and (self._wcs.wcs.spec) < 0:
+            ctypelist = list(self.wcs.wcs.ctype)
+            for n in self.substitute_spec_axis_names:
```
I think it might best to use regex, or at least perhaps only checking in lower case in order to avoid issues where the axis is defined as `linear` or `WAVELENGTH` (since currently, the casing has to be absolutely correct).

hamogu replied:
Done. (In a different line, thus github does not know this is resolved.)

nmearl commented on specutils/wcs/adapters/fitswcs_adapter.py:
```diff
@@ -71,17 +81,18 @@ def spec_axis(self):
+            for n in self.substitute_spec_axis_names:
+                if n in ctypelist:
+                    self._spec_axis = ctypelist.index(n)
+                    warnings.warn("WCS has no standard complient spectral axis, 'ctypes' might not be correct. I'm guessing that axis {} labelled '{}' might be spectral and proceed with that.".format(self._spec_axis, n), AstropyUserWarning)
```
```suggestion
                    logging.warning("WCS has a non-standards spectral axis, 'ctype's might be incorrect. Assuming the axis {} labelled '{}' is spectral and proceeding.".format(self._spec_axis, n))
```

nmearl commented on specutils/wcs/adapters/fitswcs_adapter.py:
```diff
@@ -1,8 +1,10 @@
+import warnings
```
I would suggest sticking with the regular `logging` setup that specutils already uses.

eteq commented:
For a quick example of a FITS header "in the wild" that this case might make work:
```
XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                  -64 / array data type                                
NAXIS   =                    1 / number of array dimensions                     
NAXIS1  =                  940                                                  
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               
EXTNAME = 'DATA    '           / extension name                                 
BSCALE  =                  1.0                                                  
BZERO   =                  0.0                                                  
BUNIT   = ''                                                                    
WCSAXES =                    1 / Number of coordinate axes                      
CRPIX1  =                  1.0 / Pixel coordinate of reference point            
CDELT1  =                5E-09 / Coordinate increment at reference point        
CTYPE1  = 'WAVELENGTH'         / Coordinate type code                           
CRVAL1  =                6E-07 / Coordinate value at reference point            
LATPOLE =                 90.0 / [deg] Native latitude of celestial pole        
```

hamogu commented:
@eteq: Given the major changes to all of this in #462 I'm very reluctant to add more tests and more files at this point. It will make that PR much more difficult to rebase. Also, I fail to run the tests on current master even with the most current released version of adsf and gwcs. I suspect that recent changes in specutils require the dev versions of adsf and gwcs. That makes development for an independent developer like me more difficult - I just want to update three lines of code here to read my fits file and not become a developer for all sorts of other packages that StScI is working on, too. Thus, I ask to merge this without new tests for now or to include this change in #462. Feel free to open an issue to add more tests and assign it to me. I'll do that once specutils works with the asdf and gwcs versions that I get from pypi again.

nmearl commented:
Specutils should work fine with the latest gwcs and asdf packages. I've created a new environment and ensured that all tests pass.

Just to verify: what errors are you seeing in your local test failures? What version of `pytest` and `pytest-astropy` do you have installed?

hamogu commented:
@nmearl I did some more debugging but still can't figure it out: #464

hamogu commented:
Resolved comments, added test (and updated one of the existing previous tests to actually test the warning message there, too), docs are in the class header. The loaders are not documented with a lot of text in the narrative docs; I feel that writing up a longer section on that is beyond the scope of this PR.

Ready for final review, if you ask me.

hamogu commented:
Test failures are all problems with the SDSS download. I've restarted the builds in question just so we get a nice looking green light but the SDSS test is not touched in this PR (and it passed locally a few hours ago).

nmearl commented on specutils/wcs/adapters/fitswcs_adapter.py:
```diff
@@ -71,17 +81,18 @@ def spec_axis(self):
+            for n in self.substitute_spec_axis_names:
+                if n in ctypelist:
+                    self._spec_axis = ctypelist.index(n)
+                    logging.warning("WCS has a non-standards spectral axis, 'ctype's might be incorrect. Assuming the axis {} labelled '{}' is spectral and proceeding.".format(self._spec_axis, n))
```
```suggestion
                    logging.warning("WCS has a non-standard spectral axis, 'ctype's might be incorrect. Assuming the axis {} labeled '{}' is spectral and proceeding.".format(self._spec_axis, n))
```

nmearl commented on specutils/wcs/adapters/fitswcs_adapter.py:
```diff
@@ -1,8 +1,10 @@
                          WCSSUB_LATITUDE, WCSSUB_LONGITUDE, WCSSUB_SPECTRAL,
                          WCSSUB_STOKES, InvalidSubimageSpecificationError)
 
+from astropy.utils.exceptions import AstropyUserWarning
```
```suggestion
```

nmearl commented on specutils/tests/test_loaders.py:
```diff
@@ -57,6 +63,15 @@ def test_specific_spec_axis_unit(remote_data_path):
+                                   format='wcs1d-fits')
+    logmsg = caplog.record_tuples[0]
+    assert logmsg[1] == logging.WARN
+    assert "Assuming the axis 0 labelled 'wavelength' is spectral" in logmsg[2]
```
```suggestion
    assert "Assuming the axis 0 labeled 'wavelength' is spectral" in logmsg[2]
```

nmearl commented on specutils/tests/test_loaders.py:
```diff
@@ -46,6 +47,11 @@ def test_spectrumlist_GMOSfits(remote_data_path):
 
+    logmsg = caplog.record_tuples[0]
+    assert logmsg[1] == logging.WARN
+    assert "Assuming the axis 0 labelled 'linear' is spectral" in logmsg[2]
```
```suggestion
    assert "Assuming the axis 0 labeled 'linear' is spectral" in logmsg[2]
```

hamogu commented:
@nmearl I accepted your suggestions and squashed the commits such that each commit includes one logical step.
